### PR TITLE
[ET-843] [ET-879] [ETP-333] Revert "Revert "Fix/et 843 prevent fatals inactive provider""

### PR DESCRIPTION
**Pull request by cliffordp**
[ET-843] [ET-879] [ETP-333]

Reverts moderntribe/event-tickets#1690

Merge https://github.com/moderntribe/event-tickets/pull/1693 into this branch before having QA test in-branch, then merge into G20.07

10min video from #1685 that is still relevant: https://share.getcloudapp.com/GGuokpLY
Will add more to Jira ticket for QA.

[ET-843]: https://moderntribe.atlassian.net/browse/ET-843
[ET-879]: https://moderntribe.atlassian.net/browse/ET-879
[ETP-333]: https://moderntribe.atlassian.net/browse/ETP-333

### CODE REVIEW

File: src/Tribe/Commerce/PayPal/Main.php [hunk 1]
```diff
@@ -1961,7 +1962,7 @@ public function set_tickets_view( Tribe__Tickets__Commerce__PayPal__Tickets_View
 	 * @since 4.7
 	 *
 	 * @param int|object $product
-	 * @param array $attendee
+	 * @param array|boolean $attendee
```
Comment (tr1b0t) [inline src/Tribe/Commerce/PayPal/Main.php:1965]: :no_entry_sign: Missing parameter comment

<sub>hash: 041fe05d162a498d8dfd3a14c6ddcc62</sub>

File: src/Tribe/Data_API.php [hunk 2]
```diff
@@ -186,19 +185,21 @@ public function get_event_ids( $post_id ) {
 	/**
 	 * Return Ticket Provider by Order, Product, Attendee, or Ticket ID.
 	 *
-	 * @param $post_id
+	 * @since 4.5
+	 * @since TBD Use new helper method to account for possibly inactive ticket provider.
 	 *
-	 * @return bool|object
+	 * @param int|string $post_id
```
Comment (tr1b0t) [inline src/Tribe/Data_API.php:191]: :no_entry_sign: Missing parameter comment

<sub>hash: 7988be8c1db004e005575e3b102640f2</sub>

File: src/Tribe/Data_API.php [hunk 3]
```diff
@@ -186,19 +185,21 @@ public function get_event_ids( $post_id ) {
...
+	 *
+	 * @return Tribe__Tickets__Tickets|false Ticket provider instance or False if provider is not active.
 	 */
 	public function get_ticket_provider( $post_id ) {
 		$services = $this->detect_by_id( $post_id );
 
-		// if no module class return
-		if ( empty( $services['class'] ) || ! class_exists( $services['class'] ) ) {
+		if( empty( $services['class'] ) ) {
```
Comment (tr1b0t) [inline src/Tribe/Data_API.php:198]: * :no_entry_sign: Space after opening control structure is required
* :no_entry_sign: No space before opening parenthesis is prohibited
* :no_entry_sign: Expected 1 space(s) after IF keyword; 0 found
* :no_entry_sign: There must be a space between if and the first parenthesis '('

<sub>hash: 8461048d72e3a7e22fdf5e47ab545761</sub>

File: src/Tribe/Data_API.php [hunk 4]
```diff
@@ -387,10 +391,16 @@ protected function get_attendees( $post_id, $context, $services = false ) {
 		if ( ! isset( $services['class'] ) && is_numeric( $post_id ) ) {
 			return Tribe__Tickets__Tickets::get_event_attendees( $post_id );
 		} elseif ( ! isset( $services['class'] ) && ! is_numeric( $post_id ) ) {
-			return array();
+			return [];
+		}
+
+		if( ! empty( $services['class'] ) ) {
```
Comment (tr1b0t) [inline src/Tribe/Data_API.php:397]: * :no_entry_sign: Space after opening control structure is required
* :no_entry_sign: No space before opening parenthesis is prohibited
* :no_entry_sign: Expected 1 space(s) after IF keyword; 0 found
* :no_entry_sign: There must be a space between if and the first parenthesis '('

<sub>hash: f9422216d259ed60a2d5ec92ffe3ecc4</sub>

File: src/Tribe/Data_API.php [hunk 5]
```diff
@@ -387,10 +391,16 @@ protected function get_attendees( $post_id, $context, $services = false ) {
...
+			$provider = Tribe__Tickets__Tickets::get_ticket_provider_instance( $services['class'] );
 		}
 
-		$provider = call_user_func( array( $services['class'], 'get_instance' ) );
+		if( empty( $provider ) ) {
```
Comment (tr1b0t) [inline src/Tribe/Data_API.php:401]: * :no_entry_sign: Space after opening control structure is required
* :no_entry_sign: No space before opening parenthesis is prohibited
* :no_entry_sign: Expected 1 space(s) after IF keyword; 0 found
* :no_entry_sign: There must be a space between if and the first parenthesis '('

<sub>hash: 322f922645fa53f8a8498c936feba7b2</sub>

File: src/Tribe/Editor/REST/V1/Endpoints/Single_Ticket.php [hunk 6]
```diff
@@ -202,11 +202,12 @@ public function CREATE_args() {
...
 	 *
-	 * @since 4.9
-	 * @since 4.10.9 Use customizable ticket name functions.
+	 * @since  4.9
+	 * @since  4.10.9 Use customizable ticket name functions.
+	 * @since  TBD Update detecting ticket provider to account for possibly inactive provider.
 	 *
-	 * @param WP_REST_Request $request
-	 * @param $nonce_action
+	 * @param  WP_REST_Request $request
```
Comment (tr1b0t) [inline src/Tribe/Editor/REST/V1/Endpoints/Single_Ticket.php:209]: :no_entry_sign: Missing parameter comment

<sub>hash: 3a8e3cc1ab7cf96c9d21dbe320e4b52b</sub>

File: src/Tribe/Editor/REST/V1/Endpoints/Single_Ticket.php [hunk 7]
```diff
@@ -202,11 +202,12 @@ public function CREATE_args() {
...
+	 * @param  $nonce_action
```
Comment (tr1b0t) [inline src/Tribe/Editor/REST/V1/Endpoints/Single_Ticket.php:210]: :no_entry_sign: Missing parameter name

<sub>hash: 3e6662cca79e7a2c61d0a62d9d36a6c2</sub>

File: src/Tribe/Metabox.php [hunk 8]
```diff
@@ -420,34 +415,34 @@ public function ajax_attendee_uncheckin() {
...
 	/**
-	 * Get the controls (move, delete) as a string
+	 * Get the controls (move, delete) as a string.
 	 *
 	 * @since  4.6.2
 	 *
-	 * @param  array   $post_id
-	 * @param  array   $ticket_id
-	 * @param  boolean $echo
+	 * @param int     $post_id
```
Comment (tr1b0t) [inline src/Tribe/Metabox.php:422]: :no_entry_sign: Missing parameter comment

<sub>hash: 61ab8ed32f2a10353acaaed5d08a5d3e</sub>

File: src/Tribe/Metabox.php [hunk 9]
```diff
@@ -420,34 +415,34 @@ public function ajax_attendee_uncheckin() {
...
+	 * @param int     $ticket_id
```
Comment (tr1b0t) [inline src/Tribe/Metabox.php:423]: :no_entry_sign: Missing parameter comment

<sub>hash: 8a01ec45ee9bd810bf3bd82ac33f36b7</sub>

File: src/Tribe/Metabox.php [hunk 10]
```diff
@@ -420,34 +415,34 @@ public function ajax_attendee_uncheckin() {
...
+	 * @param boolean $echo
```
Comment (tr1b0t) [inline src/Tribe/Metabox.php:424]: :no_entry_sign: Missing parameter comment

<sub>hash: d172cfbe566b7a0f2906b52492c033ab</sub>

File: src/Tribe/Promoter/Observer.php [hunk 11]
```diff
@@ -79,17 +79,17 @@ public function notify_rsvp_event( $attendee_id ) {
 	}
 
 	/**
-	 * Notify to the parent Event of the ticket
+	 * Notify to the parent Event of the ticket.
 	 *
 	 * @since 4.11.5
 	 *
-	 * @param $ticket_id int The Ticket ID where to look for the Event.
+	 * @param $ticket_id int|null The Ticket ID where to look for the Event.
```
Comment (tr1b0t) [inline src/Tribe/Promoter/Observer.php:86]: :no_entry_sign: Missing parameter name

<sub>hash: 618878ac405e318707d7112afec487be</sub>

File: src/Tribe/REST/V1/Post_Repository.php [hunk 12]
```diff
@@ -904,24 +909,25 @@ protected function build_attendee_data( array $attendee, $context = 'default' )
...
 			} else {
 				$checkin_details = false;
 			}
 		}
 
 		try {
 			$attendee_order_id = $this->get_attendee_order_id( $attendee_id, $provider );
-		} catch ( ReflectionException $e ) {
-			return array();
+		}
```
Comment (tr1b0t) [inline src/Tribe/REST/V1/Post_Repository.php:925]: :no_entry_sign: Expected 1 space after closing brace; newline found

<sub>hash: e49f7456900b3c7582a9c2abd90ba2a2</sub>

File: src/Tribe/Tickets.php [hunk 13]
```diff
@@ -918,9 +925,11 @@ public function front_end_tickets_form( $content ) {}
 		 * (it may contain the user selected currency, etc)
 		 *
 		 * @param object|int $product
+		 * @param array|boolean $attendee
```
Comment (tr1b0t) [inline src/Tribe/Tickets.php:928]: :no_entry_sign: Missing parameter comment

<sub>hash: f7f742fced59eb88eb8e5a1e15c08dc3</sub>

File: src/Tribe/Tickets.php [hunk 14]
```diff
@@ -937,11 +946,11 @@ public function supports_global_stock() {
 		}
 
 		/**
-		 * Returns instance of the child class (singleton).
+		 * Returns class instance. Child classes must overload this.
 		 *
 		 * @static
 		 *
-		 * @return mixed
+		 * @return static
```
Comment (tr1b0t) [inline src/Tribe/Tickets.php:953]: :no_entry_sign: Function return type is not void, but function has no return statement

<sub>hash: 0b456e2456a3279850f526d787e6e6bb</sub>

File: src/Tribe/Tickets_Handler.php [hunk 15]
```diff
@@ -358,77 +359,82 @@ public function get_object_connections( $object ) {
...
+		];
+
+		/**
+		 * Allow filtering the relationships so providers can add their own strings.
+		 *
+		 * @since TBD
+		 *
+		 * @param array $relationships List of relationship information for providers.
+		 */
+		 $relationships = apply_filters( 'tribe_tickets_handler_relationships', $relationships );
```
Comment (tr1b0t) [inline src/Tribe/Tickets_Handler.php:426]: :warning: Found precision alignment of 1 spaces.

<sub>hash: a5104cd5ca1f98f76f5cc2feedd86559</sub>

File: src/Tribe/Tickets_Handler.php [hunk 16]
```diff
@@ -909,10 +912,11 @@ public function has_shared_capacity( $ticket ) {
 	/**
 	 * Returns whether a given object has the correct Provider for a Post or Ticket
 	 *
-	 * @since   4.6.2
+	 * @since   4.7
+	 * @since   TBD Account for possibly inactive ticket provider and better checking for default ETP provider.
 	 *
-	 * @param   int|WP_Post  $post
-	 * @param   mixed        $provider
+	 * @param int|WP_Post                    $post
```
Comment (tr1b0t) [inline src/Tribe/Tickets_Handler.php:918]: :no_entry_sign: Missing parameter comment

<sub>hash: 66792acfa04861bc77ed3705faa4edb6</sub>

File: src/Tribe/Tickets_Handler.php [hunk 17]
```diff
@@ -909,10 +912,11 @@ public function has_shared_capacity( $ticket ) {
...
+	 * @param Tribe__Tickets__Tickets|string $provider
```
Comment (tr1b0t) [inline src/Tribe/Tickets_Handler.php:919]: :no_entry_sign: Missing parameter comment

<sub>hash: 50a9c8d0e6234d103062a60afdeaf0ab</sub>

File: src/Tribe/Tickets_Handler.php [hunk 18]
```diff
@@ -925,25 +929,37 @@ public function is_correct_provider( $post, $provider ) {
 			return false;
 		}
 
-		$provider_class = get_class( $provider );
+		$provider = Tribe__Tickets__Tickets::get_ticket_provider_instance( $provider );
+
+		if( empty( $provider ) ) {
```
Comment (tr1b0t) [inline src/Tribe/Tickets_Handler.php:934]: * :no_entry_sign: Space after opening control structure is required
* :no_entry_sign: No space before opening parenthesis is prohibited
* :no_entry_sign: Expected 1 space(s) after IF keyword; 0 found
* :no_entry_sign: There must be a space between if and the first parenthesis '('

<sub>hash: 070a305667333c8e17344fd5b6ba5d65</sub>

File: src/admin-views/editor/fieldset/price.php [hunk 19]
```diff
@@ -4,19 +4,25 @@
 	$post_id = get_the_ID();
 }
 $validation_attrs = [
-	'data-validation-error="' . esc_attr( sprintf( _x( '%s price must be greater than zero.', 'ticket price validation error', 'event-tickets' ), tribe_get_ticket_label_singular( 'ticket_price_validation_error' ) ) ) . '"'
+	'data-validation-error="' . esc_attr( sprintf(
```
Comment (tr1b0t) [inline src/admin-views/editor/fieldset/price.php:7]: :no_entry_sign: Opening parenthesis of a multi-line function call must be the last content on the line

<sub>hash: 1b3ce3ea08b3a40bda58ec1cdd4fc161</sub>

File: src/admin-views/editor/fieldset/price.php [hunk 20]
```diff
@@ -4,19 +4,25 @@
...
+		// Translators: %s: singular version of the Ticket label.
+		_x( '%s price must be greater than zero.', 'ticket price validation error', 'event-tickets' ),
+		tribe_get_ticket_label_singular( 'ticket_price_validation_error' )
+	) ) . '"'
```
Comment (tr1b0t) [inline src/admin-views/editor/fieldset/price.php:11]: * :no_entry_sign: Closing parenthesis of a multi-line function call must be on a line by itself
* :no_entry_sign: Each array item in a multi-line array declaration must end in a comma

<sub>hash: 4113586095854219768f19de81a0c66d</sub>

File: src/views/blocks/rsvp/form/form.php [hunk 21]
```diff
@@ -8,20 +8,28 @@
...
  */
-$ticket_id   = $this->get( 'ticket_id' );
-$post_id     = $this->get( 'post_id' );
-$going       = $this->get( 'going' );
-$ticket_data = tribe( 'tickets.handler' )->get_object_connections( $ticket_id );
-$event_id    = $ticket_data->event;
-$must_login  = ! is_user_logged_in() && tribe( 'tickets.rsvp' )->login_required();
+
+$ticket_id = $this->get( 'ticket_id' );
+$post_id   = $this->get( 'post_id' );
```
Comment (tr1b0t) [inline src/views/blocks/rsvp/form/form.php:21]: :no_entry_sign: Overriding WordPress globals is prohibited. Found assignment to $post_id

<sub>hash: 36a7431f3b07c273fbf489e6ad79088e</sub>

File: src/views/blocks/rsvp/form/form.php [hunk 22]
```diff
@@ -33,18 +41,18 @@
...
+			<?php $this->template( 'blocks/rsvp/form/quantity', [ 'ticket' => $ticket ] ); ?>
 		<?php endif; ?>
 	</div>
 
 	<div class="tribe-right">
 		<?php $this->template( 'blocks/rsvp/form/error' ); ?>
 
 		<?php if ( $must_login ) : ?>
-			<?php $this->template( 'blocks/rsvp/form/submit-login', array( 'event_id' => $event_id, 'going' => $going, 'ticket_id' => $ticket_id ) ); ?>
+			<?php $this->template( 'blocks/rsvp/form/submit-login', [ 'event_id' => $event_id, 'going' => $going, 'ticket_id' => $ticket_id ] ); ?>
```
Comment (tr1b0t) [inline src/views/blocks/rsvp/form/form.php:52]: :no_entry_sign: When a multi-item array uses associative keys, each value should start on a new line.

<sub>hash: 76fea9f8e7f74902d7fb51d79de4119a</sub>

File: src/views/blocks/rsvp/form/form.php [hunk 23]
```diff
@@ -33,18 +41,18 @@
...
 		<?php else : ?>
-			<?php $this->template( 'blocks/rsvp/form/details', array( 'ticket' => $ticket, 'post_id' => $post_id ) ); ?>
-			<?php $this->template( 'blocks/rsvp/form/attendee-meta', array( 'ticket' => $ticket, 'ticket_id' => $ticket_id ) ); ?>
+			<?php $this->template( 'blocks/rsvp/form/details', [ 'ticket' => $ticket, 'post_id' => $post_id ] ); ?>
```
Comment (tr1b0t) [inline src/views/blocks/rsvp/form/form.php:54]: :no_entry_sign: When a multi-item array uses associative keys, each value should start on a new line.

<sub>hash: 68ff42da452591461852449e1ada395d</sub>

File: src/views/blocks/rsvp/form/form.php [hunk 24]
```diff
@@ -33,18 +41,18 @@
...
+			<?php $this->template( 'blocks/rsvp/form/attendee-meta', [ 'ticket' => $ticket, 'ticket_id' => $ticket_id ] ); ?>
```
Comment (tr1b0t) [inline src/views/blocks/rsvp/form/form.php:55]: :no_entry_sign: When a multi-item array uses associative keys, each value should start on a new line.

<sub>hash: f4aaa80fcf7621bd87a24759e98f3a8a</sub>

File: src/views/registration-js/content.php [hunk 25]
```diff
@@ -115,33 +121,32 @@
...
+				$has_tpp = Tribe__Tickets__Commerce__PayPal__Main::ATTENDEE_OBJECT === $provider || in_array( Tribe__Tickets__Commerce__PayPal__Main::ATTENDEE_OBJECT, $providers_arr, true );
 				?>
 				<div
 					class="tribe-tickets__registration__event"
 					data-event-id="<?php echo esc_attr( $event_id ); ?>"
-					data-is-meta-up-to-date="<?php echo absint( $is_meta_up_to_date ); ?>"
+					data-is-meta-up-to-date="<?php echo absint( $this->get( 'is_meta_up_to_date' ) ); ?>"
 				>
-					<?php $this->template( 'registration/summary/content', array( 'event_id' => $event_id, 'tickets' => $tickets ) ); ?>
+					<?php $this->template( 'registration/summary/content', [ 'event_id' => $event_id, 'tickets' => $tickets ] ); ?>
```
Comment (tr1b0t) [inline src/views/registration-js/content.php:141]: :no_entry_sign: When a multi-item array uses associative keys, each value should start on a new line.

<sub>hash: d68809d409c3c8934eb07ffefe71278c</sub>

File: src/views/registration-js/content.php [hunk 26]
```diff
@@ -115,33 +121,32 @@
...
 
 					<div class="tribe-tickets__item__attendee__fields">
 
-						<?php $this->template( 'registration-js/attendees/error', array( 'event_id' => $event_id, 'tickets' => $tickets ) ); ?>
+						<?php $this->template( 'registration-js/attendees/error', [ 'event_id' => $event_id, 'tickets' => $tickets ] ); ?>
```
Comment (tr1b0t) [inline src/views/registration-js/content.php:145]: :no_entry_sign: When a multi-item array uses associative keys, each value should start on a new line.

<sub>hash: ce9b2997763079db5f753f557e0fdfd7</sub>

File: src/views/registration-js/content.php [hunk 27]
```diff
@@ -167,7 +172,7 @@ class="tribe-tickets__item__attendee__fields__form <?php echo sanitize_html_clas
 						</div>
 					</div>
 				</div>
-				<?php $this->template( 'registration-js/attendees/content', array( 'event_id' => $event_id, 'tickets' => $tickets, 'provider' => $providers[0] ) ); ?>
+				<?php $this->template( 'registration-js/attendees/content', [ 'event_id' => $event_id, 'tickets' => $tickets, 'provider' => $providers[0] ] ); ?>
```
Comment (tr1b0t) [inline src/views/registration-js/content.php:175]: :no_entry_sign: When a multi-item array uses associative keys, each value should start on a new line.

<sub>hash: 3fcc87720275c82e706e0acced1fee85</sub>

File: src/Tribe/Attendee_Registration/Shortcode.php [hunk 28]
```diff
@@ -56,7 +56,10 @@ public function maybe_enqueue_scripts() {
 	public function render() {
 		ob_start();
 
-		echo tribe( 'tickets.attendee_registration.view' )->display_attendee_registration_shortcode();
+		/** @var Tribe__Tickets__Attendee_Registration__View $view */
+		$view = tribe( 'tickets.attendee_registration.view' );
+
+		echo $view->display_attendee_registration_shortcode();
```
Comment (tr1b0t) [inline src/Tribe/Attendee_Registration/Shortcode.php:62]: * :warning: Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$view'
* :no_entry_sign: All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found '$view'.

<sub>hash: 456613cfdef6f8dcdbbe32b1c9e56aa3</sub>

File: src/Tribe/Attendee_Registration/View.php [hunk 29]
```diff
@@ -157,19 +152,14 @@ public function display_attendee_registration_shortcode() {
 		wp_localize_script(
 			'event-tickets-registration-page-scripts',
 			'TribeCurrency',
-			[
-				'formatting' => json_encode( $currency )
-			]
+			[ 'formatting' => json_encode( $currency ) ]
```
Comment (tr1b0t) [inline src/Tribe/Attendee_Registration/View.php:155]: :warning: json_encode() is discouraged. Use wp_json_encode() instead.

<sub>hash: 13de83984ffad2bcf40aebc8a78584b2</sub>

File: src/Tribe/Tickets_Handler.php [hunk 30]
```diff
@@ -807,7 +810,7 @@ public function get_post_totals( $post ) {
 			'has_unlimited' => false,
 			'has_shared'    => $global->is_enabled(),
 			'tickets'       => count( $tickets ),
-			'capacity'      => tribe_get_event_capacity( $post ),
+			'capacity'      => (int) tribe_get_event_capacity( $post ), // Could be null.
```
Comment (sc0ttkclark) [inline src/Tribe/Tickets_Handler.php:813]: This sounds out of scope of this ticket, we might want to be very wary about changes like this.

Comment (sc0ttkclark) [reply src/Tribe/Tickets_Handler.php:813]:
```suggestion
			'capacity'      => tribe_get_event_capacity( $post ),
```

Comment (cliffordp) [reply src/Tribe/Tickets_Handler.php:813]: @sc0ttkclark

iirc, https://github.com/moderntribe/event-tickets/blob/revert-1690-revert-1685-fix/ET-843-prevent-fatals-inactive-provider/src/Tribe/Admin/Columns/Tickets.php#L151 was throwing a notice and is the only spot this 'capacity' is used (via IDE autocomplete).

Searching through all our plugins for `['capacity']`, none such usages explicitly check for `null`, some for `''` and some cast to _int_ on their own.

I'd say with the thoroughness of the QA this ticket will receive, it's able to be included but obviously that's your call.

**Comment by cliffordp** [comment]
@sc0ttkclark Not sure if you're referring to the _['capacity']_ check I already commented on or another piece, but this provider checking is fundamental throughout our plugins and going through so many places in such detail surfaced some things that my provider logic change affected (e.g. which commerce tabs appear on Attendees Report) and I didn't want that to be the reason all this work got kicked back from QA (e.g. the page heading/title missing when on the TPP Orders tab when TPP isn't the default ticket provider and the default ticket provider isn't able to be active, such as WooCommerce plugin deactivated)

All that as preface for: yes, I agree and understand and try hard to keep into scope and it's not always easy and thanks for your help and and and ;)

**Pull request merged**